File: ios/mobile/BaseHaptic/Stadium/CheerSignalsLoader.swift
import Foundation
import Supabase

struct CheerSignal: Codable, Identifiable {
    let teamCode: String
    let role: String              // "home" | "away"
    let cheerText: String
    let primaryColorHex: String
    let hapticPatternId: String
    var id: String { teamCode + role }

    enum CodingKeys: String, CodingKey {
        case teamCode = "team_code"
        case role
        case cheerText = "cheer_text"
        case primaryColorHex = "primary_color_hex"
        case hapticPatternId = "haptic_pattern_id"
    }
}

struct CheerSignalEntry: Codable, Identifiable {
    let gameId: String
    let stadiumCode: String
    let fireAtIso: String
    let signals: [CheerSignal]
    var id: String { gameId }

    enum CodingKeys: String, CodingKey {
        case gameId = "game_id"
        case stadiumCode = "stadium_code"
        case fireAtIso = "fire_at_iso"
        case signals
    }
}

// 백엔드 집계는 iOS+Android 합산이며, raw 이벤트에만 platform 보존.
final class CheerSignalsLoader {
    static let shared = CheerSignalsLoader()
    private static let cacheKey = "cheer_signals_v1_cache"

    private(set) var cached: [CheerSignalEntry] = []

    private init() { load() }

    func refresh() async {
        let dateString = Self.dateFormatter.string(from: Date())
        guard let url = URL(string: "\(Self.backendBaseURL)/cheer-signals?date=\(dateString)") else { return }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return }
            let envelope = try JSONDecoder().decode(CheerSignalsEnvelope.self, from: data)
            cached = envelope.items
            UserDefaults.standard.set(try JSONEncoder().encode(envelope.items), forKey: Self.cacheKey)
        } catch {
            print("[CheerSignals] refresh failed: \(error)")
        }
    }

    func entry(forStadium stadiumCode: String, on date: Date) -> CheerSignalEntry? {
        cached.first { $0.stadiumCode == stadiumCode }
    }

    func signal(forStadium stadiumCode: String, team: Team) -> (entry: CheerSignalEntry, signal: CheerSignal)? {
        guard let entry = entry(forStadium: stadiumCode, on: Date()),
              let signal = entry.signals.first(where: { $0.teamCode == team.rawValue }) else { return nil }
        return (entry, signal)
    }

    func fetchTeamRankings(period: String) async -> [TeamCheckinRank]? {
        guard let url = URL(string: "\(Self.backendBaseURL)/rankings/teams?period=\(period)") else { return nil }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return nil }
            let envelope = try JSONDecoder().decode(TeamCheckinRankingEnvelope.self, from: data)
            return envelope.items.compactMap { item in
                guard let team = Team(rawValue: item.teamCode) else { return nil }
                return TeamCheckinRank(rank: item.rank, team: team, count: item.count)
            }
        } catch {
            print("[CheerRankings] fetch failed: \(error)")
            return nil
        }
    }

    func postCheckin(stadium: Stadium, selectedTeam: Team, game: Game?) async throws {
        let session = try await SupabaseClientProvider.client.auth.session
        guard let url = URL(string: "\(Self.backendBaseURL)/cheer-events") else { throw URLError(.badURL) }

        var payload: [String: Any] = [
            "team_code": selectedTeam.rawValue,
            "stadium_code": stadium.code,
            "client_ts": ISO8601DateFormatter().string(from: Date()),
            "lat": stadium.latitude,
            "lng": stadium.longitude,
            "accuracy_m": stadium.radiusMeters,
            "mock_location": false,
            "platform": "ios",
            "app_version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
            "is_home_team": game?.homeTeamId == selectedTeam,
            "opponent_team_code": opponentCode(for: game, selectedTeam: selectedTeam) ?? ""
        ]
        if let game {
            payload["game_id"] = game.id
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: Self.cacheKey) else { return }
        cached = (try? JSONDecoder().decode([CheerSignalEntry].self, from: data)) ?? []
    }

    private func opponentCode(for game: Game?, selectedTeam: Team) -> String? {
        guard let game else { return nil }
        if game.homeTeamId == selectedTeam { return game.awayTeamId.rawValue }
        if game.awayTeamId == selectedTeam { return game.homeTeamId.rawValue }
        return nil
    }

    private static var backendBaseURL: String {
        let base = BackendConfig.baseURL
        return base.hasSuffix("/") ? String(base.dropLast()) : base
    }

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()
}

struct TeamCheckinRank: Identifiable {
    let rank: Int
    let team: Team
    let count: Int
    var id: String { team.rawValue }
}

private struct CheerSignalsEnvelope: Decodable {
    let items: [CheerSignalEntry]
}

private struct TeamCheckinRankingEnvelope: Decodable {
    let items: [TeamCheckinRankingItem]
}

private struct TeamCheckinRankingItem: Decodable {
    let teamCode: String
    let count: Int
    let rank: Int

    enum CodingKeys: String, CodingKey {
        case teamCode = "team_code"
        case count
        case rank
    }
}

File: apps/mobile/app/src/main/java/com/basehaptic/mobile/stadium/CheerSignalsLoader.kt
package com.basehaptic.mobile.stadium

import android.content.Context
import android.util.Log
import com.basehaptic.mobile.BuildConfig
import com.basehaptic.mobile.data.model.Team
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.util.concurrent.TimeUnit

data class StadiumInfo(
    val code: String,
    val name: String,
    val homeTeamCodes: List<String>,
    val latitude: Double,
    val longitude: Double,
    val radiusMeters: Float,
    val indoor: Boolean,
)

data class CheerSignal(
    val teamCode: String,
    val role: String,
    val cheerText: String,
    val primaryColorHex: String,
    val hapticPatternId: String,
)

data class CheerSignalEntry(
    val gameId: String,
    val stadiumCode: String,
    val fireAtIso: String,
    val signals: List<CheerSignal>,
)

data class TeamCheckinRank(
    val rank: Int,
    val team: Team,
    val count: Int,
)

object CheerSignalsLoader {
    private const val TAG = "CheerSignalsLoader"
    private const val PREFS_NAME = "stadium_cheer_cache"
    private const val KEY_STADIUMS = "stadiums_json"
    private const val KEY_SIGNALS = "cheer_signals_json"

    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    suspend fun fetchStadiums(context: Context): List<StadiumInfo> = withContext(Dispatchers.IO) {
        val endpoint = "${BuildConfig.BACKEND_BASE_URL.trimEnd('/')}/stadiums"
        val body = get(endpoint)
        if (body != null) {
            context.cachePrefs().edit().putString(KEY_STADIUMS, body).apply()
        }
        parseStadiums(body ?: context.cachePrefs().getString(KEY_STADIUMS, null).orEmpty())
    }

    suspend fun fetchCheerSignals(context: Context): List<CheerSignalEntry> = withContext(Dispatchers.IO) {
        val endpoint = "${BuildConfig.BACKEND_BASE_URL.trimEnd('/')}/cheer-signals"
        val body = get(endpoint)
        if (body != null) {
            context.cachePrefs().edit().putString(KEY_SIGNALS, body).apply()
        }
        parseCheerSignals(body ?: context.cachePrefs().getString(KEY_SIGNALS, null).orEmpty())
    }

    suspend fun fetchTeamRankings(context: Context, period: String): List<TeamCheckinRank> = withContext(Dispatchers.IO) {
        val normalizedPeriod = if (period == "season") "season" else "weekly"
        val endpoint = "${BuildConfig.BACKEND_BASE_URL.trimEnd('/')}/rankings/teams?period=$normalizedPeriod"
        val body = get(endpoint) ?: return@withContext emptyList()
        parseTeamRankings(body)
    }

    suspend fun postCheckin(
        accessToken: String,
        team: Team,
        stadiumCode: String,
        gameId: String?,
        latitude: Double?,
        longitude: Double?,
        accuracyMeters: Float?,
        mockLocation: Boolean,
    ): Boolean = withContext(Dispatchers.IO) {
        if (team == Team.NONE) return@withContext false
        val payload = JSONObject().apply {
            put("team_code", team.name)
            put("stadium_code", stadiumCode)
            put("game_id", gameId)
            put("client_ts", Instant.now().toString())
            put("lat", latitude)
            put("lng", longitude)
            put("accuracy_m", accuracyMeters)
            put("mock_location", mockLocation)
            put("app_version", BuildConfig.VERSION_NAME)
            put("platform", "android")
        }
        val request = Request.Builder()
            .url("${BuildConfig.BACKEND_BASE_URL.trimEnd('/')}/cheer-events")
            .addHeader("Authorization", "Bearer $accessToken")
            .post(payload.toString().toRequestBody("application/json".toMediaType()))
            .build()
        runCatching {
            client.newCall(request).execute().use { it.isSuccessful }
        }.getOrElse {
            Log.w(TAG, "postCheckin failed", it)
            false
        }
    }

    private fun get(endpoint: String): String? {
        val request = Request.Builder().url(endpoint).get().build()
        return runCatching {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    Log.w(TAG, "GET $endpoint returned ${response.code}")
                    return null
                }
                response.body?.string()
            }
        }.getOrElse {
            Log.w(TAG, "GET $endpoint failed", it)
            null
        }
    }

    private fun parseStadiums(text: String): List<StadiumInfo> {
        if (text.isBlank()) return emptyList()
        return runCatching {
            val items = JSONObject(text).optJSONArray("items") ?: JSONArray()
            buildList {
                for (index in 0 until items.length()) {
                    val item = items.optJSONObject(index) ?: continue
                    val homeTeams = item.optJSONArray("home_team_codes") ?: JSONArray()
                    add(
                        StadiumInfo(
                            code = item.optString("code"),
                            name = item.optString("name"),
                            homeTeamCodes = buildList {
                                for (teamIndex in 0 until homeTeams.length()) {
                                    add(homeTeams.optString(teamIndex))
                                }
                            },
                            latitude = item.optDouble("latitude"),
                            longitude = item.optDouble("longitude"),
                            radiusMeters = item.optDouble("radius_meters").toFloat(),
                            indoor = item.optBoolean("indoor"),
                        )
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    private fun parseCheerSignals(text: String): List<CheerSignalEntry> {
        if (text.isBlank()) return emptyList()
        return runCatching {
            val items = JSONObject(text).optJSONArray("items") ?: JSONArray()
            buildList {
                for (index in 0 until items.length()) {
                    val item = items.optJSONObject(index) ?: continue
                    val signals = item.optJSONArray("signals") ?: JSONArray()
                    add(
                        CheerSignalEntry(
                            gameId = item.optString("game_id"),
                            stadiumCode = item.optString("stadium_code"),
                            fireAtIso = item.optString("fire_at_iso"),
                            signals = buildList {
                                for (signalIndex in 0 until signals.length()) {
                                    val signal = signals.optJSONObject(signalIndex) ?: continue
                                    add(
                                        CheerSignal(
                                            teamCode = signal.optString("team_code"),
                                            role = signal.optString("role"),
                                            cheerText = signal.optString("cheer_text"),
                                            primaryColorHex = signal.optString("primary_color_hex"),
                                            hapticPatternId = signal.optString("haptic_pattern_id"),
                                        )
                                    )
                                }
                            },
                        )
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    private fun parseTeamRankings(text: String): List<TeamCheckinRank> {
        if (text.isBlank()) return emptyList()
        return runCatching {
            val items = JSONObject(text).optJSONArray("items") ?: JSONArray()
            buildList {
                for (index in 0 until items.length()) {
                    val item = items.optJSONObject(index) ?: continue
                    val team = Team.fromString(item.optString("team_code"))
                    if (team == Team.NONE) continue
                    add(
                        TeamCheckinRank(
                            rank = item.optInt("rank", index + 1),
                            team = team,
                            count = item.optInt("count"),
                        )
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    private fun Context.cachePrefs() = applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
